libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/UnitTestTools.hs

```diff
@@ -0,0 +1,184 @@
+--   Introduces the 'fresh' operation which is 'arbitrary' for a fixed size tuple, where all the
+--   elements have the same type. Supports pattern matching against list-like fixed tuples,
+--   with no partial pattern matching. For example if you wrote this
+--   [a,b,c,d] <- arbitary @[Int]
+--      or this
+--   [a,b,c,d] <- vectorOf 3 @Int
+--   There is the possiblity of a pattern match failure. Instead we would write
+--   (a,b,c,d) <- fresh @4 @Int
+--   If the width of the tuple and the @4, doesn't match, a type error occurs.
+--   See Test.Cardano.Ledger.Shelley.UnitTests.IncrementalStakeTest for an example of its use.
+module Test.Cardano.Ledger.UnitTestTools (
+  -- * Operations
+  Fresh (fresh),
+  addr,
+  keyHashWitVKeyPairFor,
+  someKeyHashWitVKeyPair,
+  txOut,
+  rewardAccount,
+  emptyPState,
+  emptyDState,
+  emptyVState,
+  emptyCertState,
+  genLens,
+  arbitraryLens,
+  rewardCoin,
+  depositCoin,
+
+  -- * Types supported
+  Hash,
+  Addr (..),
+  RewardAccount (..),
+  Coin (..),
+  CompactForm (CompactCoin),
+  Credential (..),
+  StakeReference (..),
+  StandardCrypto,
+  Crypto,
+  DRep (DRepCredential),
+  KeyHash,
+  KeyRole (..),
+  SafeHash,
+  TxId,
+  TxIn,
+  IndividualPoolStake (IndividualPoolStake),
+  WitVKey,
+  RDPair (..),
+  UTxO (..),
+  PoolParams (..),
+  CertState (..),
+  DState (..),
+  PState (..),
+  VState (..),
+  StrictMaybe (..),
+) where
+
+import Cardano.Crypto.DSIGN.Class ()
+import Cardano.Ledger.Address (Addr (..), RewardAccount (..))
+import Cardano.Ledger.BaseTypes (Network (..), StrictMaybe (..))
+import Cardano.Ledger.CertState (CertState (..), DState (..), PState (..), VState (..))
+import Cardano.Ledger.Coin (Coin (..), CompactForm (CompactCoin))
+import Cardano.Ledger.Compactible (fromCompact)
+import Cardano.Ledger.Core
+import Cardano.Ledger.Credential (Credential (..), StakeReference (..))
+import Cardano.Ledger.Crypto (Crypto, StandardCrypto)
+import Cardano.Ledger.DRep (DRep (DRepCredential))
+import Cardano.Ledger.Keys (Hash, KeyHash (..), KeyRole (..), WitVKey (..), coerceKeyRole, hashKey)
+import Cardano.Ledger.PoolDistr (IndividualPoolStake (IndividualPoolStake))
+import Cardano.Ledger.PoolParams (PoolParams (..))
+import Cardano.Ledger.SafeHash (SafeHash)
+import Cardano.Ledger.TxIn (TxId (..), TxIn (..))
+import Cardano.Ledger.UMap (RDPair (..))
+import Cardano.Ledger.UTxO (UTxO (..))
+import Cardano.Ledger.Val (Val (..))
+import Data.Default.Class (Default (def))
+import Data.Kind (Type)
+import GHC.TypeNats (KnownNat, Nat)
+import Lens.Micro
+import Test.Cardano.Ledger.Core.Arbitrary ()
+import Test.Cardano.Ledger.Core.KeyPair (KeyPair (..), mkWitnessVKey)
+import Test.QuickCheck
+
+-- ==================================================
+
+-- | Supports patterns like
+--   (tom, john, ann, ron, mary) <- fresh @5 @(Credential 'Staking (EraCrypto era))
+class KnownNat n => Fresh (n :: Nat) (a :: Type) where
+  type Tuple n a :: Type
```

Comment:
```suggestion
  type Tuple n a :: Type | a <- n
```
I wonder if something like this would work

Reply:
The problem with type families is that they don't reduce until both arguments are known, so we can't use functional dependency information until both are known. I tried it. It didn't work, but it was a good idea.